python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll

```diff
@@ -8,4 +8,234 @@ private import semmle.python.dataflow.new.DataFlow
+
+      override DataFlow::Node getRegexNode() { result.asCfgNode() = node.getArg(0) }
+      override Attribute getRegexMethod() { result = node.getNode().getFunc().(Attribute) }
+    }
+
+    /**
+     * A call to `re.finditer`
+     * See https://docs.python.org/3/library/re.html#re.finditer
+     */
+    private class ReFindIterCall extends RegexExecution::Range, DataFlow::CfgNode {
+      override CallNode node;
+
+      ReFindIterCall() { node.getFunction() = re_attr("finditer").asCfgNode() }
+
+      override DataFlow::Node getRegexNode() { result.asCfgNode() = node.getArg(0) }
+      override Attribute getRegexMethod() { result = node.getNode().getFunc().(Attribute) }
+    }
+
+    /**
+     * A call to `re.sub`
+     * See https://docs.python.org/3/library/re.html#re.sub
+     */
+    private class ReSubCall extends RegexExecution::Range, DataFlow::CfgNode {
+      override CallNode node;
+
+      ReSubCall() { node.getFunction() = re_attr("sub").asCfgNode() }
+
+      override DataFlow::Node getRegexNode() { result.asCfgNode() = node.getArg(0) }
+      override Attribute getRegexMethod() { result = node.getNode().getFunc().(Attribute) }
+    }
+
+    /**
+     * A call to `re.subn`
+     * See https://docs.python.org/3/library/re.html#re.subn
+     */
+    private class ReSubNCall extends RegexExecution::Range, DataFlow::CfgNode {
+      override CallNode node;
+
+      ReSubNCall() { node.getFunction() = re_attr("subn").asCfgNode() }
+
+      override DataFlow::Node getRegexNode() { result.asCfgNode() = node.getArg(0) }
+      override Attribute getRegexMethod() { result = node.getNode().getFunc().(Attribute) }
+    }
+
+    /**
+     * A call to `re.compile`
+     * See https://docs.python.org/3/library/re.html#re.match
+     */
+    private class ReCompileCall extends RegexExecution::Range, DataFlow::CfgNode {
+      override CallNode node;
+
+      ReCompileCall() { node.getFunction() = re_attr("compile").asCfgNode() }
+
+      override DataFlow::Node getRegexNode() { result.asCfgNode() = node.getArg(0) }
+      override Attribute getRegexMethod() { 
+        exists (DataFlow::AttrRead reMethod |
+          reMethod = re_exec_attr() and
+          node.getFunction() = reMethod.getObject().getALocalSource().asCfgNode() and
+          result = reMethod.asExpr().(Attribute)
+        )
+       }
+    }
+
+    /** 
+     * A class for modeling expressions immediately executing a regular expression.
+     * See `re_exec_attr()`  
+    */
+    private class DirectRegex extends DataFlow::CallCfgNode, RegexExecution::Range {
+      DataFlow::Node regexNode;
+      Attribute regexMethod;
+
+      DirectRegex() {
+        // needs inheritance (?)
+        this = re_exec_attr() and regexNode = this.getRegexNode() and regexMethod = this.getRegexMethod()
+      }
+
+      override DataFlow::Node getRegexNode() { result = regexNode }
+      override Attribute getRegexMethod() { result = regexMethod }
+    }
```

Comment:
Is this actually possible? I guess using `this.getRegex(Node|Method)()` calls itself, so is it recursive? It'd be great to be able to inherit the class functions for the simplicity of the query. 

Could this class be something like this?

```codeql
DirectRegex() {
        this instanceof ReMatchCall or this instanceof ReFullMatchCall...
}
```

However I don't either know if the classes could be inherited.